controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,242 @@
+  {
+    if (is_running())
+    {
+      thread_.join();
+    }
+  }
+
+  /// Check if the async update thread is running
+  /**
+   * @return True if the async update thread is running, false otherwise
+   */
+  bool is_running() const { return thread_.joinable(); }
+
+  /// Check if the async update is triggered to stop the cycle
+  /**
+   * @return True if the async update is stopped, false otherwise
+   */
+  bool is_stopped() const
+  {
+    return (
+      async_update_stop_ ||
+      (is_initialized() &&
+       get_state_function_().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE));
+  }
+
+  /// Stops the async update thread
+  /**
+   * If the async method is running, it will notify the async thread to stop and then joins the
+   * async thread.
+   */
+  void stop_async_update()
+  {
+    if (is_running())
+    {
+      async_update_stop_ = true;
+      async_update_condition_.notify_one();
+      thread_.join();
+    }
+  }
+
+private:
+  /// Initialize the async update thread
+  /**
+   * If the async update thread is not running, it will start the async update thread.
+   * If the async update thread is already configured and running, does nothing and return
+   * immediately.
+   */
+  void initialize_async_update_thread()
+  {
+    if (!is_initialized())
+    {
+      throw std::runtime_error("AsyncFunctionHandler: need to be initialized first!");
+    }
+    if (!thread_.joinable())
+    {
+      async_update_stop_ = false;
+      trigger_in_progress_ = false;
+      async_update_return_ = T();
+      thread_ = std::thread(
+        [this]() -> void
+        {
+          // \note There might be an concurrency issue with the get_state() call here. This mightn't
+          // be critical here as the state of the controller is not expected to change during the
+          // update cycle
+          while (get_state_function_().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE &&
+                 !async_update_stop_)
+          {
+            std::unique_lock<std::mutex> lock(async_mtx_);
+            async_update_condition_.wait(
+              lock, [this] { return trigger_in_progress_ || async_update_stop_; });
+            if (async_update_stop_)
+            {
+              break;
+            }
+            async_update_return_ = async_function_(current_update_time_, current_update_period_);
+            trigger_in_progress_ = false;
+            lock.unlock();
+            async_update_condition_.notify_one();
+          }
+        });
+    }
+  }
+
+  rclcpp::Time current_update_time_;
+  rclcpp::Duration current_update_period_{0, 0};
+
+  std::function<const rclcpp_lifecycle::State &()> get_state_function_;
+  std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function_;
+
+  // Async related variables
+  std::thread thread_;
+  std::atomic_bool async_update_stop_{false};
+  std::atomic_bool trigger_in_progress_{false};
+  std::atomic<T> async_update_return_;
```

Comment:
Well this is just the return enum from the update method, they usually have OK, ERROR in case of the controllers and OK, ERROR and DEACTIVATE In case of hardware

https://github.com/saikishor/ros2_control/blob/47a72456eae7d12a277c5e5204dcc777ad55abb3/controller_interface/include/controller_interface/controller_interface_base.hpp#L32-L40

https://github.com/saikishor/ros2_control/blob/47a72456eae7d12a277c5e5204dcc777ad55abb3/hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp#L20-L27